### PR TITLE
Arclines

### PR DESCRIPTION
**Pull request by rcooke-ast**
I've translated the arclines code into PypeIt. This also involved some modifications that will (eventually) allow for us to not recompute the wavelength solution for multislit data (Alternatively, if a poor solution is found for one slit, we could use a solution from another slit to pin down the final model for all slits). This feature is not implemented yet, but should be possible to implement this before the v1.0 release.

This line of development was necessary in the build-up to having fully integrated ThAr fits, which will be in a forthcoming PR. I've decided to break this up a little so that it's not too onerous!

### CODE REVIEW

File: pypeit/core/wavecal/autoid.py [hunk 1]
```diff
@@ -0,0 +1,799 @@
...
+    Parameters
+    ----------
+    spec : spectrum
+    lines : list
+      List of arc lamps on
+    wv_cen : float
+      Guess at central wavelength
+    disp : float
+      Dispersion A/pix
+    siglev
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: can you add to this doc string?  
I guess I'm too blame for it. ;)

File: pypeit/core/wavecal/autoid.py [hunk 2]
```diff
@@ -0,0 +1,799 @@
...
+
+
+def general(spec, lines, ok_mask=None, min_ampl=300., islinelist=False,
+            outroot=None, debug=False, do_fit=True, verbose=False,
+            fit_parm=None, min_nmatch=0, lowest_ampl=200.,
+            binw=None, bind=None, nstore=10, use_unknowns=False):
+    """
+    Parameters
+    ----------
+    spec
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: This one I will 'demand' we document right.

File: pypeit/core/wavecal/autoid.py [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+
+        # Append the results to the full list
+        all_final_fit.append(final_fit.copy())
+        all_patt_dict.append(patt_dict.copy())
+
+    # Return
+    return all_patt_dict, all_final_fit
+
+
+def general_old(spec, lines, ok_mask=None, min_ampl=300., outroot=None, do_fit=True,
```
Comment (profxj) [inline pypeit/core/wavecal/autoid.py]: remove?

File: pypeit/core/wavecal/defs.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
+""" Module for definitions
```
Comment (profxj) [inline pypeit/core/wavecal/defs.py]: for arcline definitions

File: pypeit/core/wavecal/fitting.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    xfit = xfit[mask==0]
+    yfit = yfit[mask==0]
+    ions = all_idsion[ifit][mask==0]
+    # Final RMS
+    rms_ang = utils.calc_fit_rms(xfit, yfit, fit, aparm['func'],
+                                   minv=fmin, maxv=fmax)
+    rms_pix = rms_ang/disp
+    #
+    '''
+    if msgs._debug['arc']:
```
Comment (profxj) [inline pypeit/core/wavecal/fitting.py]: remove?

File: pypeit/core/wavecal/patterns.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
+""" Module for finding patterns in arc line spectra
+"""
+from __future__ import (print_function, absolute_import, division, unicode_literals)
+
+import numpy as np
+from scipy.ndimage.filters import maximum_filter
+from scipy.ndimage.morphology import generate_binary_structure, binary_erosion
+import numba as nb
+import pdb
```
Comment (profxj) [inline pypeit/core/wavecal/patterns.py]: remove

File: pypeit/core/wavecal/patterns.py [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        elif (nuni == 1) & (nmatch == 2):
+            scores.append('Risk')
+        else:
+            scores.append('Amb')
+    # Return
+    return scores
+
+
+@nb.jit(nopython=True, cache=True)
+def triangles(detlines, linelist, npixels, detsrch=5, lstsrch=10, pixtol=1.0):
```
Comment (profxj) [inline pypeit/core/wavecal/patterns.py]: add a brief description.

File: pypeit/core/wavecal/patterns.py [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+        patt_dict['scores'] = scores
+        patt_dict['IDs'] = detids
+    return
+
+
+def score_triangles(counts):
+    """  Grades for the triangle results
+    Parameters
+    ----------
+    fidx
```
Comment (profxj) [inline pypeit/core/wavecal/patterns.py]: should be ```counts```
and please add to doc str

**Comment by profxj** [comment]
Travis is  failing because a file is missing.

File: pypeit/core/wavecal/autoid.py (hunk 1 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/autoid.py]: done

File: pypeit/core/wavecal/autoid.py (hunk 2 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/autoid.py]: done - I should have remembered to do this one!

File: pypeit/core/wavecal/autoid.py (hunk 3 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/autoid.py]: oops - thanks, good catch.

File: pypeit/core/wavecal/defs.py (hunk 4 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/defs.py]: done

File: pypeit/core/wavecal/fitting.py (hunk 5 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/fitting.py]: done

File: pypeit/core/wavecal/patterns.py (hunk 6 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/patterns.py]: done

File: pypeit/core/wavecal/patterns.py (hunk 7 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/patterns.py]: Indeed - I've put a decent description here to explain the pattern matching algorithm.

File: pypeit/core/wavecal/patterns.py (hunk 8 above)
Comment (rcooke-ast) [reply pypeit/core/wavecal/patterns.py]: Thanks, and I've added to this too.

**Comment by rcooke-ast** [comment]
OK - I've finished updating now, feel free to comment and/or merge when ready.

**Comment by profxj** [comment]
Thanks so much for doing this @rcooke-ast .

Merging.

**Pull request merged**